File: nim/luhn/luhn.nim
import strutils

proc invalidCharacters(number: string): bool =
  not number.allCharsInSet(Digits + Whitespace)

proc digits(number: string): seq[int] =
  for c in number:
    if isDigit(c):
      result.add(parseInt($c))

proc toDouble(digit: int): int =
  if digit > 4:
    digit * 2 - 9
  else: 
    digit * 2

proc doubledSum(digits: seq[int]): int =
  proc indexForDouble(i: int): bool = (digits.len - i) mod 2 == 0

  for i in countdown(digits.len - 1, 0):
    if i.indexForDouble:
      result.inc(digits[i].toDouble)
    else:
      result.inc(digits[i])

proc isValid*(number: string): bool =
  if number.invalidCharacters:
    return false

  let digits = number.digits
  if digits == [0]:
    return false

  digits.doubledSum mod 10 == 0

File: zig/luhn/luhn.zig
const std = @import("std");

pub fn isValid(s: []const u8) bool {
    var checksum: u32 = 0;
    var numDigits: u32 = 0;

    for (0..s.len) |i| {
        const c = s[s.len - 1 - i];
        if (c == ' ') continue;
        if (!std.ascii.isDigit(c)) return false;

        var digit = c - '0';
        if (numDigits % 2 == 1) digit *= 2;
        if (digit > 9) digit -= 9;

        checksum += digit;
        numDigits += 1;
    }

    return numDigits > 1 and checksum % 10 == 0;
}
